File: c_cpp/base4k.c
/*   Base4K encoding for C
 *   Copyright (c) 2014 Secomba GmbH
 *   Licensed under the MIT license (https://raw.githubusercontent.com/secomba/base4k/master/LICENSE)
 *   Version: 1.0.0
 */

#include "base4k.h"

#include <stdlib.h>
#include <malloc.h>

b4kStatus base4kEncode(B4K_IN uint8_t* cData, B4K_IN_OUT uint32_t* ccData, B4K_OUT B4K_NEEDS_FREE uint16_t** cEncoded)
{
        static const unsigned int BASE_FLAG_START = 0x04000;
        static const unsigned int BASE1_START = 0x05000;
        static const unsigned int BASE_FLAG_SIZE = 0x100;
        static const unsigned int BASE1_SIZE = 0x05000;

        unsigned int offset = 0;
        unsigned int i = 0;
        unsigned int outputRunner = 0;
        
        *cEncoded = (uint16_t*)calloc(*ccData+1,sizeof(uint16_t));
        if(*cEncoded==NULL) {
                return B4K_ERROR;
        }

        for (i = 0; i < *ccData*2-2; i+=3) {
                if (i % 2 == 0)
                        offset = ((cData[i >> 1] << 4) | ((cData[(i >> 1) + 1] >> 4) & 0x0f)) & 0x0fff;
                else
                        offset = ((cData[(i - 1) >> 1] << 8) | (cData[(i + 1) >> 1] & 0xff)) & 0x0fff;

                (*cEncoded)[outputRunner++] = offset + BASE1_START;
        }

        if ((*ccData << 1) % 3 == 2) {
                offset = (cData[*ccData - 1] & 0xff) + BASE_FLAG_START;
                (*cEncoded)[outputRunner++] = offset;
        }
        else if ((*ccData << 1) % 3 == 1) {
                offset = (cData[*ccData - 1] & 0x0f) + BASE_FLAG_START;
                (*cEncoded)[outputRunner++] = offset;
        }

        (*cEncoded)[outputRunner] = '\0';
        *ccData = outputRunner;
        
        return B4K_SUCCESS;
}

b4kStatus base4KDecode(B4K_IN uint16_t* cEncoded, B4K_IN_OUT uint32_t* ccEncoded, B4K_OUT B4K_NEEDS_FREE uint8_t** cDecoded)
{
        static const unsigned int BASE_FLAG_START = 0x04000;
        static const unsigned int BASE1_START = 0x05000;
        static const unsigned int BASE_FLAG_SIZE = 0x100;
        static const unsigned int BASE1_SIZE = 0x05000;

        unsigned int code = 0;
        unsigned int prevCode = 0;
        unsigned int nrOfBytes = 0;
        unsigned int i = 0;
        unsigned int outputRunner = 0;
        unsigned int encodedLen = 0;

        if(*ccEncoded == B4K_AUTO) {
                for(;cEncoded[encodedLen] != 0; encodedLen++);
        }
        else {
                encodedLen = *ccEncoded;
        }
        *cDecoded = calloc(encodedLen*2, sizeof(uint8_t));
        if(*cDecoded==NULL)
                return B4K_ERROR;
                
        for (i = 0; i < encodedLen; i++) {
                prevCode = code;
                code = cEncoded[i];

                // check for valid encoding
                if (!(code >= BASE1_START && code < BASE1_START + BASE1_SIZE)) {
                        if (i < encodedLen - 1 || !(code >= BASE_FLAG_START && code < BASE_FLAG_START + BASE_FLAG_SIZE))
                                return B4K_ERROR;
                }

                if (code >= BASE1_START) {
                        code -= BASE1_START;
                }
                else {
                        code -= BASE_FLAG_START;
                        if (i % 2 == 0)
                                (*cDecoded)[outputRunner++] = code & 0xff;
                        else
                                (*cDecoded)[outputRunner++] = ((prevCode << 4) | (code & 0x0f)) & 0xff;
                        break;
                }
                if (i % 2 == 0) {
                        (*cDecoded)[outputRunner++] = (code >> 4) & 0xff;
                }
                else {
                        (*cDecoded)[outputRunner++] = ((prevCode << 4) | ((code & 0x0f00) >> 8)) & 0xff;
                        (*cDecoded)[outputRunner++] = code & 0xff;
                }
        }

        *ccEncoded = outputRunner;

        return B4K_SUCCESS;
}

File: java/com/secomba/base4k/Base4K.java
/*   Base4K encoding for Java
 *   Copyright (c) 2014 Secomba GmbH
 *   Licensed under the MIT license (https://raw.githubusercontent.com/secomba/base4k/master/LICENSE)
 *   Version: 1.0.0
 */

package com.secomba.base4k;

import java.util.ArrayList;

public class Base4K 
{
	private static final int BASE_FLAG_START = 0x04000;
	private static final int BASE1_START     = 0x06000;
	private static final int BASE1_START_LEGACY = 0x05000;
	private static final int BASE_FLAG_SIZE  = 0x100;
	private static final int BASE1_SIZE      = 0x01000;

	private Base4KVersion _version;

	public Base4K(Base4KVersion version) {
		_version = version;
	}
	
	public String encode(byte[] raw)
	{
		int offset = 0;
		StringBuffer result = new StringBuffer();
		
		for(int i=0;i<raw.length*2-2;i+=3) {
			if(i%2==0) {
				offset = ((raw[i/2]<<4)|((raw[i/2+1]>>>4)&0x0f))&0x0fff;
			} else { 
				offset = ((raw[(i-1)/2]<<8)|(raw[(i+1)/2]&0xff))&0x0fff;
			}

			offset += _version == Base4KVersion.V1 ? Base4K.BASE1_START_LEGACY : Base4K.BASE1_START;
			
			result.appendCodePoint(offset);
		}
		
		if((raw.length*2) % 3 == 2) {
			offset = (raw[raw.length-1]&0xff)+Base4K.BASE_FLAG_START;
			result.appendCodePoint(offset);
		} else if((raw.length*2) % 3 == 1) {
			offset = ((raw[raw.length-1]&0x0f))+Base4K.BASE_FLAG_START;
			result.appendCodePoint(offset);
		}

		return result.toString();
	}

	public byte[] decode(String encoded) throws DecodingFailedException
	{
		try {
			return decodeInternal(encoded, Base4K.BASE1_START);
		} catch (DecodingFailedException e) {
			return decodeInternal(encoded, Base4K.BASE1_START_LEGACY);
		}
	}
	
	private byte[] decodeInternal(String encoded, int base1Start) throws DecodingFailedException
	{
		int code = 0;
		ArrayList<Byte> buffer = new ArrayList<Byte>();
		ArrayList<Integer> intCollector = new ArrayList<Integer>();
		
		for(int i=0;i<encoded.length();i++) {
			code = encoded.codePointAt(i);
			
			if(!(code>=base1Start && code<base1Start+BASE1_SIZE)) {
				if(i<encoded.length()-1 || !(code>=BASE_FLAG_START && code<BASE_FLAG_START+BASE_FLAG_SIZE)) {
					throw new DecodingFailedException();
				}
			}
			intCollector.add(code);
		}
		
		for(int i=0;i<intCollector.size();i++) {
			if(intCollector.get(i)>=base1Start) {
				intCollector.set(i, intCollector.get(i) - base1Start);
			} else {
				intCollector.set(i, intCollector.get(i) - BASE_FLAG_START);
				if(i%2==0) {
					buffer.add((byte)(intCollector.get(i)&0xff));
				} else { 
					buffer.add((byte)(((intCollector.get(i-1)<<4)|((intCollector.get(i)&0x0f))&0xff)));
				}
				break;
			}
			if(i%2==0) {
				buffer.add((byte)((intCollector.get(i)>>>4)&0xff));
			} else {
				buffer.add((byte)(((intCollector.get(i-1)<<4)|((intCollector.get(i)&0x0f00)>>>8))&0xff));
				buffer.add((byte)(intCollector.get(i)&0xff));
			}
		}
			
		byte[] bResult = new byte[buffer.size()];
		for(int i=0;i<buffer.size();i++) {
			bResult[i] = buffer.get(i);
		}
		return bResult;
	}
}
